is causal hints for transformer

mikekgfb opened the pull request:
Summary:
make is_causal hint flags available for the top level transformer module.

It's debatable whether this is useful -- at present we autodetect causal masks for src and tgt masks in transformer encoder and decoder, respectively. is_causal flags available woul enable users to short-cut this check by asserting whether they mask is causal, or not.

I am putting this diff up for discussion, not as a solution.  Not doing anything may be the right solution, unless there is strong (data-driven) user demand. -- it appears the consensus is to move ahead with this, as per discussions below.

@cpuhrsch @mikaylagawarecki @jbschlosser @janEbert 

Test Plan: sandcastle

Differential Revision: D47373260

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -3831,7 +3831,11 @@ def test(encoder_input_shape, decoder_input_shape,
+                 src_is_causal=False, tgt_is_causal=False,
+                 memory_is_causal=False, raises=True):
+
+            raises=True
```
Looks like this was left here by accident

mikekgfb replied:
Actually - this is a trial balloon on removing raises as an argument and just always make it True. ;).

mikekgfb replied:
> Hm got it, in that case could we just remove it from the function signature and delete the argument at the other places where it is used

That's the plan - I just want to make sure all tests pass, before I hack up lots of call sites, and then have to redo them. Understood that this crutch can't be upstreamed as-is. (the only way to run all unit tests is github, since I'm working under fbcode... )

mikaylagawarecki replied:
gotcha! just flagging :)

mikaylagawarecki reviewed (commented):
Please fix lint

ezyang commented:
cc'ing @lezcano as we had similar API design questions for things like "is this matrix singular". IIRC our resolution was that adding these kinds of hints made sense and we should be willing to do it.

lezcano commented:
I don't remember that the specific conversation about "is the matrix singular", but I remember that sort of conversation appeared in two contexts:

[`linalg.lstsq`](https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html) has a `driver` arg that lets you choose the algorithm that will be used. We wrote in its docs a comprehensive guide on how to choose the correct `driver` arg for your application.

The other discussion we were having was about writing a set of tensor subclasses to mark that certain input matrices have some properties (e.g. they are symmetric or Hermitian, or SPD, or singular...). Then, the function would choose the most efficient algorithm available for a given set of inputs. The user would need to initially "tag" matrices with these classes, and then the algorithms could potentially return tagged matrices.
This approach is nice, because it decouples the properties of the inputs from the API, so it's more extensible. Now, it just makes sense if you expect this API to be extended.

janEbert commented on torch/nn/modules/transformer.py:
```diff
@@ -132,6 +134,28 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
+                the causal mask. Providing incorrect hints can result in
+                incorrect execution, including forward and backward
+                compatibility.
+            memory_is_causal: If specified, applies a causal mask as
+                memory mask.
```
My opinion was that cross-attention is almost never going to be causal, so it would be a waste to make the check most of the time. In contrast, the encoder self-attention may still be causal when the `TransformerEncoder` is used for a decoder-only model.

mikekgfb replied:
Are you arguing it should not even exist as a variable?  (We currently have it here, and for decoder, but only True/False, no auto detection (None) to avoid the overhead of checking, and init to `Flase`, i.e., nobody has to bother.

The question is which one is less surprising, having it (because it's not used frequently) or not having it (for consistency of _mask with _is_causal).

janEbert commented:
See also #98327.

janEbert commented on torch/nn/modules/transformer.py:
```diff
@@ -863,9 +889,9 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
-        mask: Optional[Tensor],
-        is_causal: Optional[bool],
-        size: Optional[int]
+        mask: Optional[Tensor]=None,
+        is_causal: Optional[bool]=None,
+        size: Optional[int]=None,
```
Not sure how picky you are about code style, but there should be spaces around the equals signs here.

janEbert replied:
Also, at least the `mask` argument should not be `None` by default, right? Since that one should always be supplied.

mikekgfb replied:
I think it can be because we're not checking whether the mask is None (which is a legit call to forward).  So, then we need to check this condition at all call sites, or allow None.  (Whether it should default to None is a valid ... )

janEbert commented:
I think this change makes sense to have consistency between the Transformer modules. To me, it makes sense that these flags should be available when calling a `Transformer` as well. Otherwise you lose access to these flags if you instantiate a `Transformer`; due to the defaults, users do not need to worry about them (just as before), but can still supply the flags for performance improvements.

mikaylagawarecki reviewed (commented):
Agree with the above point that this allows `nn.Transformer` to propagate these kwargs to `TransformerDecoder/TransformerEncoder` and hence `TransformerEncoderLayer/TransformerDecoderLayer` (some of which already had these kwargs in the 2.0 release (e.g. [TransformerDecoderLayer](https://pytorch.org/docs/2.0/generated/torch.nn.TransformerDecoderLayer.html?highlight=transformerdecoderlayer#torch.nn.TransformerDecoderLayer))) so we should add them for consistency

mikekgfb commented:
> Hm got it, in that case could we just remove it from the function signature and delete the argument at the other places where it is used

That's the plan - I just want to make sure all tests pass, before I hack up lots of call sites, and then have to redo them.  Understood that this crutch can't be upstreamed as-is.  (the only way to run all unit tests is github, since I'm working under fbcode... )

Commit:
is causal hints for transformer (#106143)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106143

make is_causal hint flags available for the top level transformer module.

It's debatable whether this is useful -- at present we autodetect causal masks for src and tgt masks in transformer encoder and decoder, respectively. is_causal flags available woul enable users to short-cut this check by asserting whether they mask is causal, or not.

Test Plan: sandcastle

Differential Revision: D47373260

fbshipit-source-id: eac809aacdcd8e08a7af6e0a7d0d7727431b8d7f

mikekgfb commented:
> Please fix lint

Fixed!

mikaylagawarecki reviewed (approved):
Thanks!

mikekgfb commented:
@pytorchbot merge


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: mikaylagawarecki ***@***.***>
Sent: Thursday, August 3, 2023 6:03:25 PM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] is causal hints for transformer (PR #106143)

@mikaylagawarecki approved this pull request. Thanks! — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread. Message ID: <pytorch/pytorch/pull/106143/review/1561917660@ github. com>
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

@mikaylagawarecki approved this pull request.

Thanks!

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/106143#pullrequestreview-1561917660>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHOWNOPZRVEIACVIXD3XTQN23ANCNFSM6AAAAAA22K5VYM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
